resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/internal/amid/AIMDLimiter.java

```diff
@@ -0,0 +1,219 @@
+/*
+ *
+ *  Copyright 2019: Bohdan Storozhuk, Mahmoud Romeh
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.bulkhead.adaptive.internal.amid;
+
+import static java.lang.Math.max;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.github.resilience4j.bulkhead.Bulkhead;
+import io.github.resilience4j.bulkhead.BulkheadConfig;
+import io.github.resilience4j.bulkhead.adaptive.AdaptiveBulkheadConfig;
+import io.github.resilience4j.bulkhead.adaptive.LimitPolicy;
+import io.github.resilience4j.bulkhead.adaptive.internal.config.AIMDConfig;
+import io.github.resilience4j.bulkhead.event.BulkheadLimit;
+import io.github.resilience4j.bulkhead.event.BulkheadOnLimitDecreasedEvent;
+import io.github.resilience4j.bulkhead.event.BulkheadOnLimitIncreasedEvent;
+import io.github.resilience4j.core.lang.NonNull;
+import io.github.resilience4j.core.metrics.FixedSizeSlidingWindowMetrics;
+import io.github.resilience4j.core.metrics.Metrics;
+import io.github.resilience4j.core.metrics.SlidingTimeWindowMetrics;
+import io.github.resilience4j.core.metrics.Snapshot;
+
+/**
+ * limit adapter based sliding window metrics and AMID algorithm
+ */
+public class AIMDLimiter implements LimitPolicy<Bulkhead> {
+	private static final Logger LOG = LoggerFactory.getLogger(AIMDLimiter.class);
+	private static final long MILLI_SCALE = 1_000_000L;
+	public static final String DROPPING_THE_LIMIT_WITH_NEW_MAX_CONCURRENT_CALLS = "Dropping the limit with new max concurrent calls {}";
+	private final AtomicInteger currentMaxLimit;
+	private final Consumer<BulkheadLimit> publishEventConsumer;
+	private final Metrics metrics;
+	private final AdaptiveBulkheadConfig<AIMDConfig> amidConfigAdaptiveBulkheadConfig;
+	private final long desirableLatency;
+
+
+	public AIMDLimiter(@NonNull AdaptiveBulkheadConfig<AIMDConfig> config, Consumer<BulkheadLimit> publishEventConsumer) {
+		this.amidConfigAdaptiveBulkheadConfig = config;
+		if (amidConfigAdaptiveBulkheadConfig.getConfiguration().getSlidingWindowType() == AIMDConfig.SlidingWindow.COUNT_BASED) {
+			this.metrics = new FixedSizeSlidingWindowMetrics(config.getConfiguration().getSlidingWindowSize());
```

Comment:
I think the limit policy should not care how metrics are measured. The limit policy implementation should just get some input and return some adaption value. Like a function.  The adaption of the bulkhead itself should not be part of the Limit Policy, but part of the AdaptiveBulkhead implementation.

We should move this code into the AdaptiveLimitBulkhead.